Completely replace logging infrastructure with zf_log

cbs228 opened the pull request:
The legacy logger has known issues which are summarized in #71. Some of these could lead to UB and crashes. Replace the logger with [zf_log](https://github.com/wonder-mice/zf_log). This also corrects all `printf()` format warnings reported by gcc 13. A few of these warnings were serious and could potentially lead to crashes.

Besides bugfixes, the following new functionality is introduced:

* **breaking CLI option changes**: the deprecated `-v` and `-V` verbosity switches are removed. Use `-H` host commands instead. Add a new `-m`|`--nologfile` option to suppress file output.

* **automatic rollover**: ardopcf now switches to a new log file at UTC midnight. This permits the use of [logrotate(8)](https://linux.die.net/man/8/logrotate) or other programs to remove old log files.

* **debug log decoration**: the ARDOPDebug log uses zf_log-style timestamps and includes the log level. The format is slightly different.

* **session log time**: the ARDOPSession log now uses an ISO 8601 timestamp

* **cross-platform logging**: log processing is no longer part of platform-specific source files

Now for the quirks:

* The `CONSOLELOG` and `FILELEVEL` TNC commands now use zf_log log levels instead of the legacy syslog-esque levels. This is API-BREAKING behavior for non-standard TNC commands.

* While zf_log has fewer verbosity levels than legacy ardopc, it has enough levels for our purpose. The following verbosity levels are removed:

  * `LOGEMERGENCY` was not used anywhere. Using this level now just disables logs, which is what it did before.

  * `LOGCRIT` and `LOGALERT` are mapped to `ZF_LOGF` ("fatal")

  * `LOGNOTICE` was not used anywhere

  If we want more granularity, we might prefer log "tags" for individual modules instead of trying to add more log levels.

* If ardopcf is compiled in release mode (`-O3 -DNDEBUG`), the "verbose" log level is compiled out and will never be output. This increases speed of release-mode builds at the expense of output options. We can disable this behavior if you want.

* The new logger is almost certainly not suitable for multithreaded programs. zf_log itself appears to be, but our add-on behavior should be presumed racy. We'll have to address this if we ever want to multi-thread.

This PR is huge. You can reduce the diffs by applying the following commands:

<details>
<summary>Click to view</summary>

```bash
# add zf_log from github
git subtree add --squash -P lib/zf_log \
  https://github.com/wonder-mice/zf_log.git \
  5a3f2b46c3f62b97e2b33fbdb9d391abb64644f7

# apply zf_log patch
curl -L 'https://github.com/pflarue/zf_log/commit/bcc7fe02bbd812aa88578f106b17d3fab8415c6b.patch' >/tmp/time-utc-option.patch
git am --directory=lib/zf_log /tmp/time-utc-option.patch

# add mustuse.h from #79, which you have already
# reviewed
git format-patch --stdout -1 \
  8b771d2f9103f11ac3ccd8754fb5923d6a91bfd9 \
  -- src/common/mustuse.h | \
    git am

# reproduce find and replace for new log API
find src -iname '*.c' -exec \
  sed -i -r \
    -e 's/WriteDebugLog\(\s*LOG(CRIT|ALERT)\s*,\s*/ZF_LOGF(/g' \
    -e 's/WriteDebugLog\(\s*LOGERROR\s*,\s*/ZF_LOGE(/g' \
    -e 's/WriteDebugLog\(\s*LOGWARNING\s*,\s*/ZF_LOGW(/g' \
    -e 's/WriteDebugLog\(\s*LOGINFO\s*,\s*/ZF_LOGI(/g' \
    -e 's/WriteDebugLog\(\s*LOGDEBUG\s*,\s*/ZF_LOGD(/g' \
    -e 's/WriteDebugLog\(\s*LOGDEBUGPLUS\s*,\s*/ZF_LOGV(/g' \
    -e 's/sprintf\s*\(\s*(HostCmd|strFault),\s*"/snprintf(\1, sizeof(\1), "/g' \
    \{\} \;
git commit -m 'Reproduce find and replace'
```
</details>

Closes #71

Blocks #79

cbs228 committed:
common: add ARDOP_MUSTUSE compile-time check

Add a macro which ensures—subject to compiler support—that the
return value of a function is not discarded. This politely
encourages calling functions to perform error-checking like they
should.

cbs228 committed:
build: add object sub-selection and wrapping for tests

Permit unit tests to downselect the production `$(OBJS)` they
want. This permits us to mock functions in some places but not
others. Test still depend on all object files, however.

For cmocka, we can use `-Wl,--wrap=SYMBOL` to wrap `SYMBOL` in a
new function named `__wrap_SYMBOL`.

* `SYMBOL` will resolve to `__wrap_SYMBOL`, which has an
  implementation we can define in cmocka.

* The real genuine `SYMBOL` remains accessible under
  `__real_SYMBOL`.

New Makefile logic permits individual tests to specify their own
wrap lists.

cbs228 committed:
ardop: eliminate printf() warnings

Fix format string / argument mismatch in all logging statements
and some `snprintf()` commands as well.

Some of these logic errors—especially the missing arguments—might
cause UB and a crash.

cbs228 committed:
ardop: switch from WriteDebugLog to ZF_LOG

The legacy ARDOP log levels are mapped to zf_log as follows:

* `LOGDEBUGPLUS` → `ZF_LOGV`
* `LOGDEBUG` → `ZF_LOGD`
* `LOGINFO` → `ZF_LOGI`
* `LOGWARNING` → `ZF_LOGW`
* `LOGNOTICE`: unused
* `LOGERROR` → `ZF_LOGE`
* `LOGCRIT` → `ZF_LOGF`
* `LOGALERT` → `ZF_LOGF`
* `LOGEMERGENCY`: unused

This commit is generated largely from:

```bash
find src -iname '*.c' -exec \
  sed -i -r \
    -e 's/WriteDebugLog\(\s*LOG(CRIT|ALERT)\s*,\s*/ZF_LOGF(/g' \
    -e 's/WriteDebugLog\(\s*LOGERROR\s*,\s*/ZF_LOGE(/g' \
    -e 's/WriteDebugLog\(\s*LOGWARNING\s*,\s*/ZF_LOGW(/g' \
    -e 's/WriteDebugLog\(\s*LOGINFO\s*,\s*/ZF_LOGI(/g' \
    -e 's/WriteDebugLog\(\s*LOGDEBUG\s*,\s*/ZF_LOGD(/g' \
    -e 's/WriteDebugLog\(\s*LOGDEBUGPLUS\s*,\s*/ZF_LOGV(/g' \
    -e 's/sprintf\s*\(\s*(HostCmd|strFault),\s*"/snprintf(\1, sizeof(\1), "/g' \
    \{\} \;
```

Also replace some `sprintf()` calls with safer `snprintf()`.

cbs228 committed:
ardop: adjust logging statement verbosity

The "fatal" log level is overused. Downgrade many uses of it to
other levels.

cbs228 committed:
ardop: convert Debugprintf → ZF_LOG

Convert all Debugprintf() calls to the appropriate ZF_LOG macro.
Fix format string / argument mismatch in all logging statements.

The typical log level for old `Debugprintf`s is INFO, but we use
errors and warnings as believed to be appropriate.

cbs228 committed:
arq: convert stats log to ZF_LOG

Record ARQ statistics with the new `ardop_log_session_...()`
series of functions. Logs are written to the console, debug log,
and the session log file.

cbs228 committed:
ardop: enable new logging framework

Activate the new logger with:

  * CLI arguments at startup
  * Via the host interface

The logger is no longer part of the platform-specific C files.

cbs228 committed:
Remove legacy logging functions

Remove legacy log functions that are no longer needed:

* CloseDebugLog
* CloseStatsLog
* Debugprintf
* Statsprintf
* WriteDebugLog
* WriteLog

and their globals

cbs228 committed:
ws_server: non-virtual interface for logging callbacks

Make the `ws_debug()` and `ws_error()` functions instead of
function pointers. Internally, these logging functions still
call user-defined functions set by `ws_set_debug()` and
`ws_set_error()`. The interface is *different*, however: it uses
`va_list` instead of `...`.

The advantage is that our fully-defined `ws_debug()` function
can have a declaration that includes `__printflike`:

    void __printflike(1, 2) ws_debug(const char* fmt, ...);

This allows gcc to perform printf safety checks, which are
essential to avoiding crash bugs.

The ardopcf log-receiving functions are updated with the new
interface.

Two small printf format errors are corrected.

pflarue commented:
I'm back from some time with limited Internet connectivity.  I've read through the comments on this and the prior PRs, but haven't yet look at the code.  I'll be spending some time on these in the next few days. 

So far, the only thing that concerns me is the handling of UTC date/time.  It sounds like you've implemented a workaround to make this work.  Using UTC is important to me since it allows convenient comparison of logs generated by communicating stations which when communicating over HF may often be in different time zones.  Those communicating stations may also be running different Ardop implementations.  Amateur radio operators generally use UTC for contact logging, so compatibility with those logs is also valuable.  So  I'll look closely at this.

I've found the ability to adjust the log level without restarting ardopcf to be useful to temporarily turn up the log level for testing purposes, and then return it to a lower level for normal use.  I think that this would be particularly valuable if ardopcf were to be used as a long running process.  I plan to revisit the CODEC, PLAYBACK, and CAPTURE host commands to better facilitate this mode of use while allowing reduced use of system resources while inactive.  Such changes would make using ardopcf on a headless SBC more convenient by allowing more complete control of the program via a host interface or the WebGui.  It would be nice if logging in to the computer's console wasn't required for normal use.

That said, I'm open to also retaining the -v and -V command line options if you can suggest a reason to retain them.  I have marked all options whose functionality was duplicated by the --hostcommands option as deprecated.  In the case of -v and -V, I recognized that this slightly delays the setting of these levels, but didn't see that as a significant problem.  I welcome further discussion on this topic.

I don't anticipate the lack of multithreading support or reduced level granularity will be problems.  I expect ardopcf to remain a single threaded program.  While some adjustment of level settings associated with specific log messages may be appropriate, the lack of use of some previously defined levels suggests that we should be able to accommodate a reduction in the number of available levels.

cbs228 commented:
>  Amateur radio operators generally use UTC for contact logging, so compatibility with those logs is also valuable.

I wonder if ADIF output would be useful. But that's a problem for later.

> I've found the ability to adjust the log level without restarting ardopcf to be useful to temporarily turn up the log level for testing purposes

This is certainly very useful. I was speaking more about the `DEBUGLOG` command which starts and stops all file logging. This can already be done by adjusting the log level.

> I'm open to also retaining the -v and -V command line options if you can suggest a reason to retain them

I think that verbosity options are right up there with `--help` and `--version` when it comes to standard behavior for CLI programs and daemons. Right now, the only way to prevent a log file from being written is to either use `-v 0` or to set `--logdir /dev/null`, and the latter causes a warning. I think this is a fairly compelling use-case:

* If I was going to deploy ardopcf as an RMS or other server node, I would probably opt to use systemd and just use the journal for logs. In that case, separate console logs are completely superfluous.

* Many SBCs and some desktops don't run systemd, but almost all of them support `syslog()`.

On that note, we should probably add a `--syslog` option that replaces the stdout console log with `syslog()` when enabled. zf_log has examples of this; it wouldn't be hard. WDYT? That would be a task for a future PR, not now. It would also be a Linux-only option. Alternatively, we could design a console formatter specifically for journald, but this may be less useful.

Users running `ardopcf` interactively will probably want regular logs, and regular log files are great for troubleshooting reports from end users. So they are here to stay!

cbs228 committed:
Squashed 'lib/zf_log/' content from commit 5a3f2b4

git-subtree-dir: lib/zf_log
git-subtree-split: 5a3f2b46c3f62b97e2b33fbdb9d391abb64644f7

cbs228 committed:
Merge commit '2172439c137e48d1eb4c9e4604992693c1fb1dc9' as 'lib/zf_log'

Add zf_log from

<https://github.com/wonder-mice/zf_log.git>
5a3f2b46c3f62b97e2b33fbdb9d391abb64644f7

cbs228 commented:
cf06c51b26f2a65cdf7a01ddb77c030d735e848f rebases to fix conflicts. There should be no diffs from the previous version. (See the [compare](https://github.com/pflarue/ardop/compare/27dee024e1ba804a3bf252e962aec3cb194a5f67..cf06c51b26f2a65cdf7a01ddb77c030d735e848f) button for interdiffs.)

pflarue committed:
Add option to use UTC time rather than local time

pflarue reviewed (changes requested):
Lots of good work here.  Thanks for the effort.

However, it doesn't produce UTC timestamps on Windows.  Via the comments, I've proposed an alternative that uses zf_log.c from the time-utc-option branch of https://github.com/pflarue/zf_log.  If we move forward with this, I intend to submit this change to zf_log as a pull request to to the parent repository.  Whether it is accepted there or not, I think that it is a good solution for use by ardopcf.

My comments similarly describe replacing the change to TZ and use of localtime_r() with gmtime() for a better solution in the couple of spots where a UTC date/time are used other than in zf_log.c. 

I've suggested some changes to zf_log format descriptors and removal of `_UTC` from the log file name.

A few other minor changes.

cbs228 commented:
> However, it doesn't produce UTC timestamps on Windows. Via the comments, I've proposed an alternative that uses zf_log.c from the time-utc-option branch of https://github.com/pflarue/zf_log.

I think your patch is a great alternative implementation, but I would like to investigate this issue a bit more first.

I suspect that the Windows API function [`GetLocalTime()`](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlocaltime) from [zf_log.c:799](https://github.com/wonder-mice/zf_log/blob/5a3f2b46c3f62b97e2b33fbdb9d391abb64644f7/zf_log/zf_log.c#L799) does not respect the value of `TZ`. If it does, the Windows documentation makes no mention of this. I believe that Windows doesn't really have the notion of a "process-specific time zone."

Do the log filename timestamps work correctly? Are they UTC? Is it only the zf_log timestamps which are wrong? This old [sample program](https://sourceforge.net/p/mingw/mailman/message/15863872/) from 2005 suggests that MinGW's CRT should handle `TZ` "the POSIX way."

If it's only zf_log that is wrong, we can override `g_time_cb` to inject our own version of this function which uses the MinGW `localtime_r()`. If that works, there will be no need to patch zf_log. The result won't be usable with MSVC, but we've already committed to supporting MinGW only.

pflarue commented:
I compiled ardopcf on Windows using the the code from cbs228:feature/zf_log.  By adjusting my windows system clock to late in the day so that the UTC date does not match my local date, I confirmed that while the timestamp values written to the log are in local time, the date values used for the filename of the log file are in UTC.

While this observation may allow a different solution than the one I proposed, I still think that my solution is a good choice.  Since we want UTC, using functions that provide it directly seems more straight forward than setting a process specific TZ value and then using functions intended to provide local time values.  If you feel strongly that we should avoid modifying zf_log, please explain why.  If zf_log was still being actively updated, and my changes were not accepted into the main code base, then I could understand such reluctance since it might complicate keeping a local copy of zf_log up to date with future improvements to zf_log.  However, the last change to zf_log was five years ago, so this doesn't seem to be relevant.

I agree that ardopcf is committed to only supporting MinGW for windows builds.  Doing so has allowed some simplification of the code base while providing a simple method for windows users to build from source should they wish to do so.  Thus, I continue to believe that this was a good choice.  Based on downloads of the last release of ardopcf, most users are running Linux, but I'm committed to continuing to also support Windows users.

cbs228 commented:
> the timestamp values written to the log are in local time, the date values used for the filename of the log file are in UTC

Thanks for testing this. I think this confirms my suspicions about `GetLocalTime()`.

> Since we want UTC, using functions that provide it directly seems more straight forward than setting a process specific TZ value and then using functions intended to provide local time values.

I prefer UTC as well, but I also want to avoid forcing a timezone or other locale behavior on the user. Using `localtime` and friends permit us to make the timezone runtime-selectable. This MR does not add such an option, but it leaves the door open for us to do so.

Food for thought:

* js8call uses only UTC, everywhere
* fldigi links both localtime and gmtime, but it appears to use only UTC as far as I can determine
* xastir displays only local time. xastir is mostly intended for VHF as opposed to HF
* pat logs only in local time

If you'd like to force UTC for everyone, everywhere, I'll go ahead and use your patch. If not, have a look at b8d09be642b48006b2a69e741f6d8118349e5168, which should fix the problem on Windows. Let me know which one of these you prefer and it'll be in the next revision.

There's only one other item from the review that I think needs more discussion. w.r.t. the host interface changes,

> While I'm OK with changing the level to what is probably a reasonable value, I'd prefer to also send an error to the host program.

To me, the presence of FAULT implies that the command did not complete. I'm not sure it's a good idea to both issue a FAULT **and** do something else instead (i.e., set a different log level). More importantly, I think that introducing an error message for previously acceptable input should itself be considered breaking behavior.

I can find no mention of `CONSOLELOG` or `LOGLEVEL` in the [2017 Host Interface Spec](https://www.winlink.org/sites/default/files/downloads/ardop_tnc_host_mode_interface_spec.pdf). We *could* elect to treat these as non-standard commands which are subject to revision across minor, rather than major, releases. Then perhaps the cleaner thing to do would be to just realign the  `LOGLEVEL` numerics with zf_log's logging levels. This is still breaking, but I think it's a less "hacky" approach. If we adopt the zf_log levels, we can just remove `ardop_log_verbosity_to_level()` and the legacy log levels altogether. WDYT?

pflarue commented:
I don't see the ability to choose between UTC and local time as a feature with much real value.  So, I'm not interested in making design choices to accommodate it as a possible future feature.  Let's stay with only UTC.

I also don't particularly like the "hacky" conversion between ardop's traditional log levels and those used by zf_log.  When I saw this in the pull request, I also considered proposing that we simply adopt the zf_log level values natively.  Before doing so, I looked to see whether any host programs were actually using these.  I found no use of LOGLEVEL or CONSOLELOG in Pat.  While arim and garim don't use them internally, they have a `tnc-init-cmd` configuration option to allow users to send arbitrary host commands to ardop on startup.  Using this to set ardop's LOGLEVEL was specifically recommended in some announcements for these programs.  So, I decided to accept it as written.

However, following your last comment, I reviewed where I those references to LOGLEVEL.  I found that it was specifically suggested for use with TNC-Pi9k6.  So, it probably has limited usage.  Given all that, I think that adopting the zf_log levels natively as parameters to the LOGLEVEL and CONSOLELOG host commands is a good idea.  With this change, let's also adopt your approach of setting them to the min or max valid level if a value outside the available range is given.  However, a response of the form "LOGLEVEL now 4" should always be sent to the host indicating that the change has been made.  This is currently missing.

For most users, this change will only require adjustment to the command line usage, which I've already moved toward with my deprecation of some arguments.

cbs228 reviewed (commented):
This is ready for re-review. You'll probably get a few more email notifications for other threads because the github review tool simply can't cope with this many diffs.

cbs228 commented:
Ooops, also had to fix the deprecation warning for the CLI. This should now be ready.

pflarue commented:
The handling of -v (--verboselog) and -V (--verboseconsole) do not match the description of these options defined in HelpScreen[].  They are described as providing incremental changes, not absolute values.  Since these were previously deprecated, I propose that we eliminate these two options now rather than modify them.  You previously advocated not deprecating these options.  Do you still feel that they are useful?

Whether these are removed or not, HelpScreen[] needs to be updated.  If retained, then the range of values for the LOGLEVEL and CONSOLELOG in the deprecation messages need to be adjusted.

pflarue commented:
I think this is close to ready. 

See comments regarding a couple of minor edits, either eliminating or correcting the -v and -V options, and returning a Syntax error if the LOGLEVEL or CONSOLELOG commands are given with a non-numerical value.

cbs228 commented:
> You previously advocated not deprecating these options. Do you still feel that they are useful?

There are scenarios where it is undesirable for ardopcf to write log files. When ardopcf is running as a system service under systemd or SysV init, there is probably some other logging mechanism available—i.e., syslog. In this case, the user might want a log file, or they might just want syslog to take care of it. At present, the only way to disable log files is via `--verboselog`. Some log messages are written before the `-H` commands are processed. This forces the creation of a log file.

I am considering adding a Linux-only `--syslog` option that will completely replace the `log_callback()` with one that only uses syslog only. (No log files.) The output will be more tailored for syslog and/or systemd. This will be the subject of a follow-up MR.

We could also replace `-v` and `-V` with a `--no-log-file` option that starts with `ArdopLogFileEnabled = false`. `ArdopLogFileEnabled` is controlled by the standard `DEBUGLOG` host command, so a `DEBUGLOG` could turn it on again. That might be small enough to implement here in this MR, or we could defer it. Or not do it at all. My main goal is `--syslog`.

> I think this is close to ready.

I'll get the other things taken care of. If you're okay with the general concept of `--syslog`, I'll also remove `-v` and `-V`.

pflarue commented:
Adding a --no-log-file option seems like a reasonable addition while removing  -v and -V.  Please also include a suitable single letter argument form to go with it and add a description to HelpScreen.

cbs228 committed:
common: add try_parse_long() wrapper around strtol()

Add a convenience wrapper around `strtol(3)` which parses a number
from a string. Our method returns true if the *entire* string is
valid as a number.

A new unit test file is added for `ARDOPCommon.c`.

cbs228 committed:
log: add logging core and ArdopLogFile

Introduce a new module, `log.c`, which wraps zf_log and supports
multiple output streams.

Legacy ARDOP had four outputs:

1. Console log

2. A "debug" log, which nominally mirrors the console
   output with a different verbosity level

3. A "session" log, which summarizes each ARQ session

4. An "exception" log, which is no longer used

The new log implementation provides the first three. ZF_LOG
macros should now be used instead of `WriteDebugLog`,
`Debugprintf`, or vanilla `printf()` (for the most part).

The `ArdopLogFile` object permits us to track and manage the
lifetime of multiple logging files. The new object permits
*automatic rollover* of log files. The logger will automatically
switch to the next file when the UTC day changes.

The session log now uses ISO 8601 timestamps (with nanoseconds)
and is written via the `ardop_log_session` series of functions.
Its format is otherwise unchanged.

The "verbose" log level is compiled out if NDEBUG is defined.

The unit test demonstrates mocking in action. All tests now
execute in the UTC time zone.

cbs228 committed:
common: remove verbosity opts, replace with --nologfile

Remove the `-v` and `-V` verbosity options, which have been
deprecated since v1.0.4.1.2 [1]. They are replaced with the
`--nologfile` option, which disables log files by default. Log
files may be enabled later with the `DEBUGLOG` host command.

This is a CLI-BREAKING change.

[1]: 501bc5e13114ab523eb4592a6bea1bb644fe0950

cbs228 commented:
978f6f6253fb..9d1a6c9066 should take care of all this. Ready for re-review.

* I have opted for `--nologfile` instead of `--no-log-file` for consistency with other options. No other options use middle-dashes.

* The short option for `--nologfile` is `-m`. There are only a handful of alphabet letters that we have not already used. I wanted to avoid clashes with uppercase options since sometimes upper vs. lowercase implies a negation or at least a relationship.

* In addition to fixing error-handling for `CONSOLELOG` and `LOGLEVEL`, I have also fixed `DEBUGLOG`.

* 32d053e1293c844dd33a3d19311c9054b1bdf7c8 adds a convenience wrapper around `strtol()` and some test-cases for it.

pflarue commented:
Found a couple of minor issues relating to DoTrueFalseCmd() with DEBUGLOG host command.

Also a compiler warning that is only produced when compiling for 32-bit Linux or Windows.

cbs228 commented:
9d1a6c9066d4afb8d66e12eb3f204cb28c4f7c62..d86258776747d0ff68cb5ef2c0a53bad3a73aa32 ought to take care of this.

pflarue committed:
remove declaration of strFault[] from ProcessCommandFromHost()

strFault[] was declared both as a global variable and as a local variable within ProcessCommandFromHost().  Thus, setting its value within DoTrueFalseCmd() did not have the intended effect within ProcessCommandFromHost().

pflarue commented:
Thanks for your effort on this.

I found only one remaining problem with this PR.  strFault[] was declared both as a global variable and as a local variable within ProcessCommandFromHost().  Thus, setting its value within DoTrueFalseCmd() did not have the intended effect within ProcessCommandFromHost().  So, I edited your feature/zf_log branch to eliminate the declaration within ProcessCommandFromHost() prior to merging this PR.  I believe that this fixes the problem.  If you'd prefer a different solution, feel free to submit an additional pull request.

cbs228 commented:
> strFault[] was declared both as a global variable and as a local variable within ProcessCommandFromHost()

Yes, that's the problem I found. I was actually going to do the opposite and just dispatch the FAULT message directly from `DoTrueFalseCmd()`, but your solution works too. Thanks for helping with this!
